fix(lsinitrd, dracut-initramfs-restore): detect initrd for BLS Type #1 entries

aafeijoo-suse opened the pull request:
dracut is making wrong assumptions with BLS Type #1 entries.

- The ESP is not always used to store boot loader entries, there may be also an Extended Boot Loader Partition (XBOOTLDR). The spec defines the concept of the `$BOOT` partition placeholder, as the primary place to put boot menu entry resources into.

- The `machine-id` is not always used as subdirectory, it can be the `entry-token`.

- Also, although the name "initrd" is referenced in the examples, its use is not mandatory. E.g., in openSUSE we add a suffix with its hash value.

Manually implementing this autodetection logic can be complex, but all the necessary information is provided by `bootctl`, so use it if available.

More info: https://uapi-group.org/specifications/specs/boot_loader_specification

## Checklist
- [ ] I have tested it locally
- [ ] I have reviewed and updated any documentation if relevant
- [ ] I am providing new code and test(s) for it

dalto8 commented on lsinitrd.sh:
```diff
@@ -109,6 +109,18 @@ find_initrd_for_kernel_version() {
+        base_path=$(bootctl -x)
+        # get initrd key of the selected bootloader entry (i.e., the one that
+        # is actually used to boot the system)
+        mapfile -t files < <(bootctl --json=pretty list 2> /dev/null | jq -r '.[] | select( .isSelected ) | .initrd[]' 2> /dev/null)
```
I was using systemd-boot.  I see this on both Arch and EndeavourOS.

```
[
        {
                "type" : "type1",
                "source" : "esp",
                "id" : "fc6662aae8a245a0a773eb6825b6e9d6-6.12.43-1-lts.conf",
                "path" : "/efi/loader/entries/fc6662aae8a245a0a773eb6825b6e9d6-6.12.43-1-lts.conf",
                "root" : "/efi",
                "title" : "EndeavourOS",
                "showTitle" : "EndeavourOS (6.12.43-1-lts)",
                "sortKey" : "endeavouros-6.12.43-1-lts",
                "version" : "6.12.43-1-lts",
                "machineId" : "fc6662aae8a245a0a773eb6825b6e9d6",
                "options" : "nvme_load=YES nowatchdog rw rd.luks.uuid=edd8ce92-0c1d-4eee-953e-d063cf9c0be6 root=/dev/mapper/luks-edd8ce92-0c1d-4eee-953e-d063cf9c0be6 rd.luks.uuid=5a20ac1e-eecd-4e8b-b7a7-e7092288bb3b resume=/dev/mapper/luks-5a20ac1e-eecd-4e8b-b7a7-e7092288bb3b systemd.ma>
                "linux" : "/fc6662aae8a245a0a773eb6825b6e9d6/6.12.43-1-lts/linux",
                "initrd" : [
                        "/fc6662aae8a245a0a773eb6825b6e9d6/6.12.43-1-lts/initrd"
                ],
                "isReported" : false,
                "isDefault" : true,
                "isSelected" : false,
                "addons" : null,
                "cmdline" : "nvme_load=YES nowatchdog rw rd.luks.uuid=edd8ce92-0c1d-4eee-953e-d063cf9c0be6 root=/dev/mapper/luks-edd8ce92-0c1d-4eee-953e-d063cf9c0be6 rd.luks.uuid=5a20ac1e-eecd-4e8b-b7a7-e7092288bb3b resume=/dev/mapper/luks-5a20ac1e-eecd-4e8b-b7a7-e7092288bb3b systemd.ma>
        },
        {
                "type" : "type1",
                "source" : "esp",
                "id" : "fc6662aae8a245a0a773eb6825b6e9d6-6.12.43-1-lts-fallback.conf",
                "path" : "/efi/loader/entries/fc6662aae8a245a0a773eb6825b6e9d6-6.12.43-1-lts-fallback.conf",
                "root" : "/efi",
                "title" : "EndeavourOS",
                "showTitle" : "EndeavourOS (6.12.43-1-lts-fallback)",
                "sortKey" : "endeavouros-6.12.43-1-lts-fallback",
                "version" : "6.12.43-1-lts-fallback",
                "machineId" : "fc6662aae8a245a0a773eb6825b6e9d6",
                "options" : "nvme_load=YES nowatchdog rw rd.luks.uuid=edd8ce92-0c1d-4eee-953e-d063cf9c0be6 root=/dev/mapper/luks-edd8ce92-0c1d-4eee-953e-d063cf9c0be6 rd.luks.uuid=5a20ac1e-eecd-4e8b-b7a7-e7092288bb3b resume=/dev/mapper/luks-5a20ac1e-eecd-4e8b-b7a7-e7092288bb3b systemd.ma>
                "linux" : "/fc6662aae8a245a0a773eb6825b6e9d6/6.12.43-1-lts/linux",
                "initrd" : [
                        "/fc6662aae8a245a0a773eb6825b6e9d6/6.12.43-1-lts/initrd-fallback"
                ],
                "isReported" : false,
                "isDefault" : false,
                "isSelected" : false,
                "addons" : null,
                "cmdline" : "nvme_load=YES nowatchdog rw rd.luks.uuid=edd8ce92-0c1d-4eee-953e-d063cf9c0be6 root=/dev/mapper/luks-edd8ce92-0c1d-4eee-953e-d063cf9c0be6 rd.luks.uuid=5a20ac1e-eecd-4e8b-b7a7-e7092288bb3b resume=/dev/mapper/luks-5a20ac1e-eecd-4e8b-b7a7-e7092288bb3b systemd.ma>
        },
        {
                "type" : "type1",
                "source" : "esp",
                "id" : "fc6662aae8a245a0a773eb6825b6e9d6-6.16.1-arch1-1.conf",
                "path" : "/efi/loader/entries/fc6662aae8a245a0a773eb6825b6e9d6-6.16.1-arch1-1.conf",
                "root" : "/efi",
                "title" : "EndeavourOS",
                "showTitle" : "EndeavourOS (6.16.1-arch1-1)",
                "sortKey" : "endeavouros-6.16.1-arch1-1",
                "version" : "6.16.1-arch1-1",
                "machineId" : "fc6662aae8a245a0a773eb6825b6e9d6",
                "options" : "nvme_load=YES nowatchdog rw rd.luks.uuid=edd8ce92-0c1d-4eee-953e-d063cf9c0be6 root=/dev/mapper/luks-edd8ce92-0c1d-4eee-953e-d063cf9c0be6 rd.luks.uuid=5a20ac1e-eecd-4e8b-b7a7-e7092288bb3b resume=/dev/mapper/luks-5a20ac1e-eecd-4e8b-b7a7-e7092288bb3b systemd.ma>
                "linux" : "/fc6662aae8a245a0a773eb6825b6e9d6/6.16.1-arch1-1/linux",
                "initrd" : [
                        "/fc6662aae8a245a0a773eb6825b6e9d6/6.16.1-arch1-1/initrd"
                ],
                "isReported" : false,
                "isDefault" : false,
                "isSelected" : false,
                "addons" : null,
                "cmdline" : "nvme_load=YES nowatchdog rw rd.luks.uuid=edd8ce92-0c1d-4eee-953e-d063cf9c0be6 root=/dev/mapper/luks-edd8ce92-0c1d-4eee-953e-d063cf9c0be6 rd.luks.uuid=5a20ac1e-eecd-4e8b-b7a7-e7092288bb3b resume=/dev/mapper/luks-5a20ac1e-eecd-4e8b-b7a7-e7092288bb3b systemd.ma>
        },
        {
                "type" : "type1",
                "source" : "esp",
                "id" : "fc6662aae8a245a0a773eb6825b6e9d6-6.16.1-arch1-1-fallback.conf",
                "path" : "/efi/loader/entries/fc6662aae8a245a0a773eb6825b6e9d6-6.16.1-arch1-1-fallback.conf",
                "root" : "/efi",
                "title" : "EndeavourOS",
                "showTitle" : "EndeavourOS (6.16.1-arch1-1-fallback)",
                "sortKey" : "endeavouros-6.16.1-arch1-1-fallback",
                "version" : "6.16.1-arch1-1-fallback",
                "machineId" : "fc6662aae8a245a0a773eb6825b6e9d6",
                "options" : "nvme_load=YES nowatchdog rw rd.luks.uuid=edd8ce92-0c1d-4eee-953e-d063cf9c0be6 root=/dev/mapper/luks-edd8ce92-0c1d-4eee-953e-d063cf9c0be6 rd.luks.uuid=5a20ac1e-eecd-4e8b-b7a7-e7092288bb3b resume=/dev/mapper/luks-5a20ac1e-eecd-4e8b-b7a7-e7092288bb3b systemd.ma>
                "linux" : "/fc6662aae8a245a0a773eb6825b6e9d6/6.16.1-arch1-1/linux",
                "initrd" : [
                        "/fc6662aae8a245a0a773eb6825b6e9d6/6.16.1-arch1-1/initrd-fallback"
                ],
                "isReported" : false,
                "isDefault" : false,
                "isSelected" : false,
                "addons" : null,
                "cmdline" : "nvme_load=YES nowatchdog rw rd.luks.uuid=edd8ce92-0c1d-4eee-953e-d063cf9c0be6 root=/dev/mapper/luks-edd8ce92-0c1d-4eee-953e-d063cf9c0be6 rd.luks.uuid=5a20ac1e-eecd-4e8b-b7a7-e7092288bb3b resume=/dev/mapper/luks-5a20ac1e-eecd-4e8b-b7a7-e7092288bb3b systemd.ma>
        },
        {
                "type" : "loader",
                "source" : "esp",
                "id" : "fc6662aae8a245a0a773eb6825b6e9d6-6.16.0-arch2-1.conf",
                "path" : "/sys/firmware/efi/efivars/LoaderEntries-4a67b082-0a4c-41cf-b6c7-440b29bb8c4f",
                "showTitle" : "fc6662aae8a245a0a773eb6825b6e9d6-6.16.0-arch2-1.conf",
                "isReported" : true,
                "isDefault" : false,
                "isSelected" : true,
                "addons" : null
        },
        {
                "type" : "loader",
                "source" : "esp",
                "id" : "fc6662aae8a245a0a773eb6825b6e9d6-6.16.0-arch2-1-fallback.conf",
                "path" : "/sys/firmware/efi/efivars/LoaderEntries-4a67b082-0a4c-41cf-b6c7-440b29bb8c4f",
                "showTitle" : "fc6662aae8a245a0a773eb6825b6e9d6-6.16.0-arch2-1-fallback.conf",
                "isReported" : true,
                "isDefault" : false,
                "isSelected" : false,
                "addons" : null
        },
        {
                "type" : "auto",
                "source" : "esp",
                "id" : "auto-reboot-to-firmware-setup",
                "path" : "/sys/firmware/efi/efivars/LoaderEntries-4a67b082-0a4c-41cf-b6c7-440b29bb8c4f",
                "title" : "Reboot Into Firmware Interface",
                "showTitle" : "Reboot Into Firmware Interface",
                "isReported" : true,
                "isDefault" : false,
                "isSelected" : false,
                "addons" : null
        }
]
```

aafeijoo-suse replied:
Ok, now I understand. You're not booting a BLS Type #1 entry (defined `<ESP>/loader/entries/*.conf`, which is what this patch tries to detect), you're booting an entry from the LoaderEntries EFI variable. Is the initrd self-contained? is it like an UKI?

aafeijoo-suse replied:
> It is actually. The actual entry it is booting is `/efi/loader/entries/fc6662aae8a245a0a773eb6825b6e9d6-6.16.1-arch1-1.conf`
> 
> You can see this if you look at the entry with the matching `id`.

hmm...

```
{
        "type" : "loader",
        ...
        "id" : "fc6662aae8a245a0a773eb6825b6e9d6-6.16.0-arch2-1.conf",
        ...
},
```

 `fc6662aae8a245a0a773eb6825b6e9d6-6.16.0-arch2-1.conf` is not `fc6662aae8a245a0a773eb6825b6e9d6-6.16.1-arch1-1.conf`, there is not any other matching id on the list.

aafeijoo-suse replied:
> When boot loader entries are defined through the Boot Loader Specification files, the identifier should be derived directly from the file name, **but with the .conf (Type #1 snippets) or .efi (Type #2 images) suffix removed**.

Vogtinator replied:
Ok, so you hit precisely what I described in https://github.com/dracut-ng/dracut-ng/pull/1596#issuecomment-3214265309: used initrd vs. next initrd. In your case, the used initrd no longer exists, so with the current choice in the PR it will fail by design.

Vogtinator replied:
> for the exitrd, there is no doubt that we have to unpack the same initrd used to boot the system

If the initrd is regenerated with similar contents, it will work. It's also how the code has worked previously FWICT. IMO that's better than just failing completely.

Vogtinator replied:
> And, we could add a fallback to check the "default" entry in case there is no "selected". Opinions?

I'm in favor, but it has to make sure it actually fits the running kernel. Maybe it needs to iterate entries that use the booted kernel and pick the best one...

Vogtinator commented:
I was actually looking into implementing this for `lsinitrd` myself, but hit similar issues as you. I'm also not sure what `lsinitrd` is meant to do by default: Print info about the initrd that was used for booting or the initrd that will be used for booting next? With the non BLS code, both could happen, but with BLS there's an explicit choice (`select(.isSelected)` vs. `select(.isDefault)`.

aafeijoo-suse committed:
fix(lsinitrd, dracut-initramfs-restore): detect initrd for BLS Type #1 entries

dracut is making wrong assumptions with BLS Type #1 entries.

- The ESP is not always used to store boot loader entries, there may be also an
Extended Boot Loader Partition (XBOOTLDR). The spec defines the concept of the
`$BOOT` partition placeholder, as the primary place to put boot menu entry
resources into.

- The `machine-id` is not always used as subdirectory, it can be the
`entry-token`.

- Also, although the name "initrd" is referenced in the examples, its use is not
mandatory. E.g., in openSUSE we add a suffix with its hash value.

Manually implementing this autodetection logic can be complex, but all the
necessary information is provided by `bootctl`, so use it if available.

More info: https://uapi-group.org/specifications/specs/boot_loader_specification

aafeijoo-suse commented:
PR stalled for 6 months. I'll keep a downstream patch, closing.
